doc: Amend notes on benchmarking (#31690)
bitcoin/bitcoin: doc/benchmarking.md

> The output will look similar to:

Reviewer: might want to clarify that CLang produces a very different output

Reviewer: See the examples in e.g. https://github.com/bitcoin/bitcoin/pull/31682:

> C++ compiler .......................... AppleClang 16.0.0.16000026

|            ns/block |             block/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|          361,535.93 |            2,765.98 |    0.5% |     11.02 | `CheckBlockBench`

> C++ compiler .......................... GNU 13.3.0

|            ns/block |             block/s |    err% |       ins/block |       cyc/block |    IPC |      bra/block |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|        1,096,261.84 |              912.19 |    0.1% |    7,963,390.88 |    3,487,375.26 |  2.283 |   1,266,941.00 |    1.8% |     11.03 | `CheckBlockBench`

Reviewer: > That is due to __linux__ not being set on macOS

Thanks for clarifying

> so I'll just drop this

I think it did make sense to replace the base58 examples with more representative ones

Reviewer: I've mentioned this since it confused me at the beginning, seeing base58 benchmarks as examples - which is why I worked on optimizing it at first.

Notes
---------------------

Benchmarks help with monitoring for performance regressions and act as a scope
for future performance improvements. They should cover components that impact
performance critical functions of the system. Functions are performance
critical, if their performance impacts users and the cost associated with a
degradation in performance is high. A non-exhaustive list:

> - Initial block download (Cost: slow IBD results in full node operation being

Reviewer: Users often complain about slow RPC calls - could we add those to this list?

Reviewer: Could we clarify when we should use `benchmark::PriorityLevel::HIGH` and when the priority should be lower?

Reviewer: I'd just like to avoid reviers claiming that they reject a PR because it's not in this list - but you're right, the "non-exhaustive list" should make that clear - you can resolve these

  less accessible)
- Block template creation (Cost: slow block template creation may result in
  lower fee revenue for miners)
- Block propagation (Cost: slow block propagation may increase the rate of

>   orphaned blocks)

Reviewer: very useful list - could you add explicit benchmark examples for further understanding?

Benchmark additions and performance improvements are valuable if they target
the user facing functions mentioned above. If a clear end-to-end performance
improvement cannot be demonstrated, the pull request is likely to be rejected.

> The change might also be rejected if the code bloat or review/maintenance is
> too much to justify the improvement.

Reviewer: may not be obvious what "review is too much" means:
```suggestion
The change might also be rejected if the code bloat or review/maintenance burden is
too high to justify the improvement.
```

> Benchmarks are ill-suited for testing of denial of service issues as they are

Reviewer: Seems simpler to remove the leading "of":
```suggestion
Benchmarks are ill-suited for testing denial-of-service issues, as they are
```

restricted to the same input set (introducing bias). [Fuzz

> tests](/doc/fuzzing.md) are better suited for this purpose, as they are much
> better (and aimed) at exploring the possible input space.

Reviewer: ```suggestion
tests](/doc/fuzzing.md) are better suited for this purpose, as they are specifically
aimed at exploring the possible input space.
```

Notes
---------------------

Benchmarks help with monitoring for performance regressions and act as a scope
for future performance improvements. They should cover components that impact
performance critical functions of the system. Functions are performance

> critical, if their performance impacts users and the cost associated with a

Reviewer 2: ```suggestion
critical if their performance impacts users and the cost associated with a
```

degradation in performance is high. A non-exhaustive list:

- Initial block download (Cost: slow IBD results in full node operation being
  less accessible)
- Block template creation (Cost: slow block template creation may result in

>   lower fee revenue for miners)

Reviewer 2: ```suggestion
  lower fee revenue for miners and mining centralization)
```

- Block propagation (Cost: slow block propagation may increase the rate of
  orphaned blocks)

Benchmark additions and performance improvements are valuable if they target
the user facing functions mentioned above. If a clear end-to-end performance

> improvement cannot be demonstrated, the pull request is likely to be rejected.

Reviewer 3: I also tend to lean toward a per-PR approach, but do see the value in providing insight so new contributors know to provide a more complete case for a PR. Maybe something more encouraging could accomplish a similar goal, e.g. `Supporting the case for a performance improvement with benchmark results is encouraged`.

> improvement cannot be demonstrated, the pull request is likely to be rejected.

Reviewer: It's not obvious to everyone. This is a useful change.
Some of the overly harsh/opinionated parts could indeed be dropped/reformulated, but it's important that newcomers know the basic directions.
